bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientHandler.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.xmppclient.handler;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.openhab.binding.xmppclient.action.XMPPActions;
+import org.openhab.binding.xmppclient.internal.XMPPClient;
+import org.openhab.binding.xmppclient.handler.XMPPClientConfiguration;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link XMPPClientHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Pavel Gololobov - Initial contribution
+ */
+public class XMPPClientHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(XMPPClientHandler.class);
+
+    private XMPPClient xmppClient;
+
+    private XMPPClientConfiguration config;
+
+    final Map<ChannelUID, PublishTriggerChannel> channelStateByChannelUID = new HashMap<>();
+
+    public XMPPClientHandler(Bridge thing) {
+        super(thing);
+    }
+
+    public XMPPClient getXMPPClient() {
+        return xmppClient;
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(XMPPActions.class);
+    }
+
+    @Override
+    protected void triggerChannel(ChannelUID channelUID, String event) {
+        super.triggerChannel(channelUID, event);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // not supported
+    }
+
+    @Override
+    public void initialize() {
+        logger.info("XMPP Client Start initializing");
+        config = getConfigAs(XMPPClientConfiguration.class);
+
+        xmppClient = new XMPPClient();
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        xmppClient.connect(config.host, config.port, config.username, config.password, "openHAB");
```

Comment:
Is this returning fast?

Reply:
Usually the connection takes 1-2 seconds. Is it fast or not?

Reply:
Network communication happens either asynchronously or in a dedicated thread to not block the main openHAB event loop. If there is an async method, please use that if possible.

Reply:
While Smack provides an async API, the method handling the connection state, namely `connect()` and `disconnect()` are currently blocking. Since you never want to block a main event loop, this probably means that you have to perform them in an extra thread.

Reply:
Ok. So please use something like `scheduler.schedule(this::doConnect)` and move the initialization from here to the end to a new method `doConnect()`. Then it also makes sense to set the thing status  to UNKNOWN here and to ONLINE after successful connect.

Reply:
Done. See latest commit changes.